### PR TITLE
*: add quorum package

### PR DESCRIPTION
**Pull request by BusyJay**
Adds quorum package and majority configuration. Configuration in tracker
is also updated. The quorum package is ported from etcd master.

### CODE REVIEW

**Comment by BusyJay** [comment]
@Fullstop000 Do you have time to review this? You can use command "LGTM" to give your approval.

File: src/quorum/majority.rs [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+    /// Eg. If the matched indexes are [2,2,2,4,5], it will return 2.
+    /// If the matched indexes and groups are `[(1, 1), (2, 2), (3, 2)]`, it will return 1.
+    pub fn committed_index(&self, use_group_commit: bool, l: &impl AckedIndexer) -> (u64, bool) {
+        if self.voters.is_empty() {
+            // This plays well with joint quorums which, when one half is the zero
+            // MajorityConfig, should behave like the other half.
+            return (u64::MAX, false);
+        }
+
+        let mut stack_arr: [MaybeUninit<Index>; 7] = unsafe { MaybeUninit::uninit().assume_init() };
```
Comment (hicqu) [inline src/quorum/majority.rs:49]: I prefer `SmallVec` for this.

Comment (BusyJay) [reply src/quorum/majority.rs:49]: I didn't use `SmallVec` on purpose. On the one hand, it's such a small piece of code that doesn't need to introduce another dependency; on the other hand, it has less condition branch as every operation is inlined and optimized when using array.

File: src/tracker.rs [hunk 2]
```diff
@@ -199,7 +165,7 @@ impl ProgressSet {
         let pr = Progress::new(next_idx, max_inflight);
         for id in &meta.conf_state.as_ref().unwrap().voters {
             self.progress.insert(*id, pr.clone());
-            self.configuration.voters.insert(*id);
+            self.configuration.voters.voters.insert(*id);
```
Comment (Fullstop000) [inline src/tracker.rs:168]: This looks wired... Is there potential extra fields in `Configuration` besides `voters` ?

File: src/quorum.rs [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+}
+
+impl Debug for Index {
+    #[inline]
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        Display::fmt(self, f)
+    }
+}
+
+pub trait AckedIndexer {
```
Comment (Fullstop000) [inline src/quorum.rs:45]: It could be better to add some comments for this public trait

File: src/tracker.rs (hunk 2 above)
Comment (BusyJay) [reply src/tracker.rs:168]: Yes. It will be added when Joint configuration is ported.

File: src/quorum.rs (hunk 3 above)
Comment (BusyJay) [reply src/quorum.rs:45]: It's not a public trait. It can only be accessed by the crate. We have `#![deny(missing_docs)]` check.

**Comment by Fullstop000** [comment]
LGTM

**Pull request merged**